backend/migrations/2023-07-22-110000_shadings/up.sql

```diff
@@ -0,0 +1,94 @@
+-- Your SQL goes here
+CREATE TABLE shadings (
+    id uuid PRIMARY KEY,
+    layer_id integer NOT NULL,
+    shade shade NOT NULL,
+    geometry GEOMETRY (POLYGON, 4326) NOT NULL,
+    add_date date,
+    remove_date date,
+    FOREIGN KEY (layer_id) REFERENCES layers (id) ON DELETE CASCADE
+);
+
+CREATE FUNCTION check_shade_layer_type() RETURNS trigger
+LANGUAGE plpgsql
+AS $$
+BEGIN
+    IF (SELECT type FROM layers WHERE id = NEW.layer_id) != 'shade' THEN
+        RAISE EXCEPTION 'Layer type must be "shade"';
+    END IF;
+    RETURN NEW;
+END;
+$$;
+
+CREATE TRIGGER check_shade_layer_type_before_insert_or_update
+BEFORE INSERT OR UPDATE ON shadings
+FOR EACH ROW EXECUTE FUNCTION check_shade_layer_type();
+
+-- Calculate a score for a certain position.
+-- p_layer_ids[1] ... plant layer
+-- p_layer_ids[2] ... shade layer
+CREATE OR REPLACE FUNCTION calculate_score(
+    p_map_id integer,
+    p_layer_ids integer [],
+    p_plant_id integer,
+    x_pos integer,
+    y_pos integer
+)
+RETURNS real AS $$
+DECLARE
+    plant_relation RECORD;
+    distance REAL;
+    weight REAL;
+    score REAL := 0;
+BEGIN
+    score := 0.5
+        + calculate_score_from_relations(p_layer_ids[1], p_plant_id, x_pos, y_pos)
+        + calculate_score_from_shadings(p_layer_ids[2], p_plant_id, x_pos, y_pos);
+
+    RETURN score;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calculate a score between -0.5 and 0.5 using the shadings.
+CREATE FUNCTION calculate_score_from_shadings(
+    p_layer_id integer, p_plant_id integer, x_pos integer, y_pos integer
+)
+RETURNS real AS $$
+DECLARE
+    point GEOMETRY;
+    plant_shade shade;
+    shading_shade shade;
+    all_values shade[];
+    pos1 INTEGER;
+    pos2 INTEGER;
+BEGIN
+    -- Get the preferred shade of the plant
+    SELECT shade INTO plant_shade
+    FROM plants
+    WHERE id = p_plant_id;
+
+    -- Create a point from x_pos and y_pos
+    point := ST_SetSRID(ST_MakePoint(x_pos, y_pos), 4326);
+    -- Select the shading with the darkest shade that intersects the point
+    SELECT shade INTO shading_shade
+    FROM shadings
+    WHERE layer_id = p_layer_id AND ST_Intersects(geometry, point)
+    ORDER BY shade DESC
+    LIMIT 1;
+
+    -- If there's no shading, return 0 (meaning shadings do not affect the score)
```

Comment:
This is not feasible given the little time left. I would have to rewrite the whole heatmap calculation including the backend.
As far as I understand this would require two values to be returned to the backend, one score values for red-green (same as it currently is) and an additional value for the alpha channel (there is nothing like this in place).
I can create a left over issue for that.

Reply:
*I'll give it a quick try. If its not to difficult I'll include it.

Reply:
Thx! Actually, I don't think two values would be needed, as the red is static? Having two values probably would be more flexible for future extensions, though.

Reply:
Done.